param.go

```diff
@@ -0,0 +1,72 @@
+package dig
+
+import (
+	"errors"
+	"reflect"
+)
+
+// Param represents an argument to invoked function.
+type Param struct {
```

Comment:
I guess it can be named `Optional` instead, but we can extend the behaviour(e.g. with validation, etc) and rely only on the field.